### ISSUE TITLE
Merging URIs

### ISSUE THREAD

**Issue by aadcg** [fukamachi/quri#47]
I'm wondering about the output of 
```
(quri:merge-uris "/" "https://en.wikipedia.org/wiki/URL")
```

I expected it to be `#<QURI.URI.HTTP:URI-HTTPS https://en.wikipedia.org/>`, instead of `#<QURI.URI.HTTP:URI-HTTPS https://en.wikipedia.org/wiki/URL>`.

Isn't my expectation justified, or am I misinterpreting something?

**Comment by jmercouris** [comment]
I do not believe your expectation is justified. I think at the most you could have expected "https://en.wikipedia.org/wiki/" as the output. I think what you would need to do in this case is something like this:

```
(let ((uri (quri:uri "https://en.wikipedia.org/wiki/URL")))
           (quri:make-uri :scheme (quri:uri-scheme uri)
                                  :host (quri:uri-host uri)))
```

and include everything but the path.

**Comment by jmercouris** [comment]
Note that when merging URLs there is an order, the specificity is implied by this. It is NOT order agnostic.

**Comment by aadcg** [comment]
> It is NOT order agnostic.

Indeed! This only makes me feel more certain of my proposition! 

```
(quri:merge-uris "https://en.wikipedia.org/foo" "/")
#<QURI.URI.HTTP:URI-HTTPS https://en.wikipedia.org/foo>

(quri:merge-uris "/" "https://en.wikipedia.org/foo")
#<QURI.URI.HTTP:URI-HTTPS https://en.wikipedia.org/foo>
```

I agree with the former output, disagree with the latter. 

Notice that the first argument is the REFERENCE and the latter is the BASE.

**Comment by Ambrevar** [comment]
I just read https://tools.ietf.org/html/rfc2396 section 5 and if you
look at the example in https://tools.ietf.org/html/rfc2396#appendix-C
you'll find this:

- base URL: http://a/b/c/d;p?q
- reference:  /g

Spec result: http://a/g
Quri result: http://a/g

So far so good.

If you remove the "g" from the reference, then Quri returns the base URL.

The specs say:

> 2) If the path component is empty and the scheme, authority, and
>      query components are undefined, then it is a reference to the
>      current document and we are done.

Now the question is: does "/" refer to an empty path or not?
I believe @aadcg is right and it's a bug in Quri, but I'm not completely
sure.

@fukamachi Any idea about this?

**Comment by fukamachi** [comment]
I think an empty path means "" (not "/").
In that sense, the behavior @aadcg mentioned is a bug.

**Comment by jathd** [comment]
I also think "/" is not an empty path. RFC3986 (which obsoletes RFC2396, btw), in section 4.2. (Relative Reference)
distinguishes between `path-absolute` and `path-empty`, where `path-absolute` begins with a `/`.

The bug is not in `merge-uris`, but before that, in the parsing of URIs: `(uri "/")` returns a URI with empty path, so `merge-uris` never even sees the "/". See #49.

**Comment by aadcg** [comment]
Commit fb489b9 introduced a change that breaks backwards compatibility. 

The expected behavior would be:

```lisp
(merge-uris "/" (uri "https://en.wikipedia.org/wiki/URL"))
#<URI-HTTPS https://en.wikipedia.org/>
```

In reality we get:

```lisp
(merge-uris "/" (uri "https://en.wikipedia.org/wiki/URL"))
#<URI https://en.wikipedia.org/>
```

Notice the difference in the type.

**Comment by Ambrevar** [comment]
Good catch!  I'll work on it.

**Comment by aadcg** [comment]
Actually I'm already trying to fix it. I might take a bit longer than you but I'd like to try it!

What do you think?

**Comment by Ambrevar** [comment]
New fiew merged with #58.
